# stana
                                           
[![R-CMD-check](https://github.com/noriakis/stana/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/noriakis/stana/actions/workflows/R-CMD-check.yaml) [![CodeFactor](https://www.codefactor.io/repository/github/noriakis/stana/badge)](https://www.codefactor.io/repository/github/noriakis/stana)

Strain-level metagenomic analysis in R. Import and analyse, plot output of the software like [MIDAS](https://github.com/snayfach/MIDAS), [MIDAS2](https://github.com/czbiohub/MIDAS2), [metaSNV v1, metaSNV v2](https://github.com/metasnv-tool/metaSNV) and [inStrain](https://github.com/MrOlm/inStrain). Will grow into the complete package. Primarly developed for `MIDAS` and `MIDAS2`.

## Installation
```r
devtools::install_github("noriakis/stana")
```

## Documentation

The documentation is available using `pkgdown` at [https://noriakis.github.io/software/stana_pkgdown](https://noriakis.github.io/software/stana_pkgdown). The detailed usage is available [here](https://noriakis.github.io/software/stana), using `bookdown`.

## Pipeline

<img src="https://github.com/noriakis/software/blob/main/images/stana_pipeline.png?raw=true" width="800px">

## Interactive inspection of metagenotyping results based on clinical variables

<img src="https://github.com/noriakis/software/blob/main/images/stana_app.png?raw=true" width="800px">