---
title: "stana"
author: "Noriaki Sato"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
vignette: >
  %\VignetteIndexEntry{stana}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.width=5,
                      fig.height=5,
                      warning=FALSE,
                      message=FALSE)
```

# stana

`stana` is a library for strain-level analysis of metagenomic dataset using single nucleotide variants profile and gene contents.
Multiple softwares are described, such as `MIDAS2`, `metaSNV`, and `InStrain`.
The library load the profile to `stana` class object and perform various analysis like marker gene detection and visualization of trees based on metadata.
Here, we included the sample stana class object profiled from 26 metagenome samples processed with `MIDAS2` against UHGG, and random grouping from `PRJEB9584`. The data was loaded with `loadMIDAS2` function. One species profile, identifier `100003` (`s__Alistipes shahii`) is loaded.

```{r load, eval=FALSE}
## Specify the output of `merge` command with grouping
stana <- loadMIDAS2("merge_midas2_sample", cl=randomGroup, candSp="100003",
                    loadSummary=TRUE,loadInfo=TRUE,loadDepth=TRUE,
                    taxtbl=taxtbl, db="uhgg")
```

When loading the dataset, filter options can be applied. `filtType` specifies how the filtering should be applied, `whole` or `group`. `filtNum` specifies the number, `filtPer` specifies fraction of samples. For instance, If `group` and `filtNum=2`, the species with sampled number above or equal 2 per group are included.

```{r load_data}
library(stana)
## Examine sample object
load(system.file("extdata", "sysdata.rda", package = "stana"))
stana
```

## Filtering

The species that satisfies the filtering criteria specified can be accessed by the slot `clearSnps` storing species ID. If `taxtbl` was provided on `load*` functions, `clearSnpsSpecies` has species name.

```{r filt}
getSlot(stana, "clearSnps")
getSlot(stana, "clearSnpsSpecies")
```

## Calling consensus sequence

You can call the consensus sequence based on SNV frequency table. The implementation and filtering options are based on the original `MIDAS` script (`call_consensus.py`). The resulting FASTA will be written to the current directory, and is loaded into `fastaList` slot.

```{r cons, eval=TRUE, fig.width=5, fig.height=5}
library(phangorn)
library(ggtree)
stana <- consensusSeq(stana, species="100003")

## Tree estimation and visualization by `phangorn` and `ggtree`
getFasta(stana)
stana <- inferAndPlotTree(stana)
getTree(stana)
```

## Perform PERMANOVA based on the groups

```{r permanova, eval=TRUE}
stana <- doAdonis(stana, specs = "100003",target="tree")
getSlot(stana, "adonisList")
```

## Compare gene abundances

```{r wilcox}
gn <- compareGenes(stana, "100003")
gn[[1]]
```

## Identify key elements for differentiating groups

Using the profiled matrices, the `Boruta` algorithm can be run to identify the key elements that distinguish the groups. The `doBoruta` function performs the `Boruta` algorithm between the clusters specified in `cl` (if `NULL`, the `cl` slot is used). If pre-processed matrices are to be used, please pass them to `mat`.

```{r boruta, eval=TRUE, fig.width=5, fig.height=5}
library(Boruta)

## You can provide preprocessed matrix to `mat`
doBoruta(stana, target = "genes", sp="100003")
```

## Parsing functional annotations

Functional annotation of genes can be obtained through `PATRIC` server (used in `MIDAS`), and annotation results such as ortholog assignment using `eggNOG-mapper v2`. `checkPATRIC` and `checkEGGNOG` function can be used to parse the obtained data. The below example returns KEGG ORTHOLOGY from `eggNOG-mapper v2` annotation file. Here, the annotation file is large and the code is not evaluated.

```{r egg, eval=FALSE}
## Not to be evaluated
annotationFile <- "100003_uhgg_eggnog_out.emapper.annotations"
annot <- checkEGGNOG(annotationFile, ret = "KEGG_ko")
annot
```

Using the parsed annotation tibble, gene copy number data frame can be summarised by user-defined function, such as `mean`.
Note that annotation tibble must have `ID` and `value` column specifying gene ID in `ID` and functional annotation such as KEGG ORTHOLOGY ID in `value`.

```{r merge, eval=FALSE}
## Not to be evaluated
summariseAbundance(stana, "100224", annot)
```

## Visualization

Several visualization functions are implemented.

### Plotting mean coverage across samples

```{r cov}
plotCoverage(stana, "100003") ## grouping is already set
```

### Plotting gene abundances
```{r genes, fig.width=8}
plotGenes(stana, "100003",
    geneID=getSlot(stana, "genes")[[1]] |> row.names() |> head(3))
```



```{r sess}
sessionInfo()
```